### PR TITLE
Documentation and Cleanup for Release

### PR DESCRIPTION
**Pull request by idavis**

### CODE REVIEW

File: eng/psakefile.ps1 [hunk 1]
```diff
@@ -150,6 +150,7 @@ function Install-LlvmFromSource {
     $Env:PKG_NAME = Get-PackageName
     $Env:CMAKE_INSTALL_PREFIX = $packagePath
     $Env:INSTALL_LLVM_PACKAGE = $true
+    Assert $false -failureMessage "TODO: Migration in progress"
```
Comment (troelsfr) [inline eng/psakefile.ps1:153]: Is the TODO recorded in the issues?

File: examples/generator/bernstein_vazirani.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    """
+
+    def __init__(self, qubit_count: int = 8, pattern: int = 12):
+        """
+        :param qubit_count: number of qubits to use
+        :type qubit_count: int
+        :param pattern: integer representation of the bitstring pattern
+        :type pattern: int
+        """
+        self.builder = QirBuilder("Bernstein-Vazirani")
```
Comment (troelsfr) [inline examples/generator/bernstein_vazirani.py:25]: Given that you annotate types, is it worthwhile to annotate the class members as well? I.e. 
```python
class BernsteinVazirani:
    """
    Implementation of the Bernstein-Vazirani quantum algorithm
    """
    builder: QirBuilder
    
    # ...
```

File: examples/jit/bernstein_vazirani.ll [hunk 3]
```diff
@@ -0,0 +1,536 @@
+; ModuleID = 'Bernstein-Vazirani'
+source_filename = "./bernstein_vazirani.ll"
```
Comment (troelsfr) [inline examples/jit/bernstein_vazirani.ll:2]: Is this file needed or is it an output from PyQIR that could be reproduced without checking the `ll` file in?

File: pyqir-generator/README.md [hunk 4]
```diff
@@ -1,22 +1,102 @@
...
+    is one of the Bell states: | x 〉| y 〉 = 1/sqrt(2) [|0〉|0〉 + |1〉|1〉]
+    If "false", the two qubits are entangled such that the state is:
+    | x 〉| y 〉 = 1/4 [|0〉|0〉 + |0〉|1〉 + |1〉|0〉 + |1〉|1〉]
+"""
+entangle: bool = True
 
-```shell
-pip install maturin
+builder = QirBuilder("Bell")
+builder.add_quantum_register("qubit", 2)
```
Comment (troelsfr) [inline pyqir-generator/README.md:31]: My understanding may be wrong, but I believe this function allocates/generates/reserves 2 quantum registers with names `qubit0` and `qubit1`? If so, I would suggest that the rename is made plural `add_quantum_registers`. One could optionally make an `add_quantum_register("qubit")` which would register `qubit` with no numbers appended.

File: pyqir-jit/pyqir_jit/gatelogger.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+# Licensed under the MIT License.
+
+from pyqir_jit import GateSet
+
+
+class GateLogger(GateSet):
+    def __init__(self):
+        # call parent class constructor
+        super().__init__()
+        self.number_of_registers = 0
```
Comment (troelsfr) [inline pyqir-jit/pyqir_jit/gatelogger.py:11]: Same as above: May be worthwhile to annotate types of the class members.

File: pyqir-jit/pyqir_jit/jit.py [hunk 6]
```diff
@@ -4,14 +4,14 @@
 from typing import Any
 from .pyqir_jit import *
```
Comment (troelsfr) [inline pyqir-jit/pyqir_jit/jit.py:5]: Nit & optional: Import the items that are used in the code instead of the full library. Reason: If you have two or more of such statements, it becomes substantially harder to know where classes/functions live.

**Review by troelsfr — APPROVED** [review]
Nice and clean code. Only minor suggestions for improvements.

Note: I have not read the Rust code in great detail as I do not feel qualified for it. I would suggest a second reviewer for this part of the code.

File: eng/psakefile.ps1 (hunk 1 above)
Comment (idavis) [reply eng/psakefile.ps1:153]: It is being tracked and worked on in the background.

File: examples/jit/bernstein_vazirani.ll (hunk 3 above)
Comment (idavis) [reply examples/jit/bernstein_vazirani.ll:2]: @troelsfr It is needed in order for the example to work on its own. We didn't want the samples for one project to depend on running another.

File: pyqir-jit/README.md [hunk 7]
```diff
@@ -1,22 +1,75 @@
...
+See [Building](../docs/building.md)
 
-Alternatively, install tox and run the tests inside an isolated environment:
+## Limitations
 
-```shell
-tox -e py
-```
\ No newline at end of file
+- QIR entrypoint for JIT must be named `QuantumApplication__Run`
```
Comment (bettinaheim) [inline pyqir-jit/README.md:74]: What happens if the entry point takes arguments?
What happens if there is a branching based on a measurement result?

File: pyqir-jit/README.md [hunk 8]
```diff
@@ -1,22 +1,75 @@
 # pyqir-jit
 
-## Building and Testing
+## Examples
+
+There are JIT examples in the repository:
+- [Bernstein-Vazirani](../examples/jit/bernstein_vazirani.py) ([Bernstein-Vazirani QIR](../examples/jit/bernstein_vazirani.ll))
```
Comment (bettinaheim) [inline pyqir-jit/README.md:6]: both links currently give me a 404. I would also suggest to merge this with the sentence below.

File: pyqir-jit/README.md [hunk 9]
```diff
@@ -1,22 +1,75 @@
...
+
+Let's walk though the example. Suppose we have the generated IR (`bernstein_vazirani.ll`). We can evaluate it with the `NonadaptiveJit`, and `GateLogger` to print out a simple log of the quantum application.
```
Comment (bettinaheim) [inline pyqir-jit/README.md:8]: E.g.
```suggestion
Let's look at how to log the gate sequence for the following example:
- [Bernstein-Vazirani](../examples/jit/bernstein_vazirani.py)
 We can evaluate the [generated QIR](../examples/jit/bernstein_vazirani.ll) with the `NonadaptiveJit`, and `GateLogger` to print out a simple log of the quantum application.
```

File: pyqir-jit/README.md [hunk 10]
```diff
@@ -1,22 +1,75 @@
 # pyqir-jit
```
Comment (bettinaheim) [inline pyqir-jit/README.md:2]:
```suggestion

The 'pyqir_jit' package provides an easy way to execute generated QIR for the purpose of
1. easily testing and experimenting with QIR code
2. connecting it to low-level Python-based lab software such as e.g. [QCoDes.](https://qcodes.github.io/Qcodes/user/intro.html) 

It contains the necessary [just-in-time compilation](https://en.wikipedia.org/wiki/Just-in-time_compilation) infrastructure as well an extensibility mechanism to define what actions to perform when a gate is applied in Python.
```

File: pyqir-jit/README.md [hunk 11]
```diff
@@ -1,22 +1,75 @@
...
```
Comment (bettinaheim) [inline pyqir-jit/README.md:1]: For consistency with the other readme:
```suggestion
# pyqir_jit
```

File: pyqir-parser/README.md [hunk 12]
```diff
@@ -1,22 +1,13 @@
 # pyqir-parser
```
Comment (bettinaheim) [inline pyqir-parser/README.md:1]: For consistency with the other readme:
```suggestion
# pyqir_parser
```

File: examples/jit/bernstein_vazirani.ll [hunk 13]
```diff
@@ -0,0 +1,536 @@
+; ModuleID = 'Bernstein-Vazirani'
```
Comment (bettinaheim) [inline examples/jit/bernstein_vazirani.ll:1]: I would suggest to check this in as bitcode instead, and make sure it contains the necessary qirlib pieces in it if needed.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -1,9 +1,11 @@
 name: CI
 on:
   push:
-    branches: [ $default-branch ]
+    branches:
+    - main
```
[the remark this answers is not in the document]
Comment (idavis) [reply .github/workflows/ci.yml:5]: When a PR is created targeting main, the feature branches will trigger builds. Do you want builds triggered before the PR is created?

File: pyqir-jit/pyqir_jit/jit.py [hunk 15]
```diff
@@ -4,14 +4,14 @@
...
 
-class QirJit(object):
-    """
-    The QirJit object loads bitcode/QIR for evaluation and processing
 
+class NonadaptiveJit(object):
```
[the remark this answers is not in the document]
Comment (idavis) [reply pyqir-jit/pyqir_jit/jit.py:8]: Nonadaptive : not able to adapt to the environment, can't make decisions based on results.

File: eng/psakefile.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (idavis) [reply eng/psakefile.ps1:153]: It is where the LLVM builds need to be slotted in. I've been working on them in the background and the cleanup work didn't depend on them.

File: examples/jit/bernstein_vazirani.ll (hunk 13 above)
Comment (idavis) [reply examples/jit/bernstein_vazirani.ll]: Done.

File: docs/compatibility.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+PyQIR runs on most x86-64 operating systems that can run Python 3.6+; however, not all of these systems are equally compatible. Operating systems are grouped into tiers that represent the level of compatibility users can expect.
+
+* Tier 1 systems are compatible. For tier 1 systems:
+  * operating system is used in automated tests
+  * installation packages provided for them
+* Tier 2 systems should be compatible with PyQIR and can be used relatively easily. For tier 2 systems:
+  * informal testing may have been done
+  * the packages for tier 1 systems will likely work in tier 2 systems
+
+### Tier 1
```
[the remark this answers is not in the document]
Comment (idavis) [reply docs/compatibility.md:14]: I was hoping to add these post merge when we've uploaded a release. I can point to the general releases page for now, though it will be empty.

File: pyqir-jit/README.md (hunk 8 above)
Comment (idavis) [reply pyqir-jit/README.md:11]: There is a mismatch of path resolution between local and on GH. This doc will be moving to a new file to cleanup support for pypi.

File: pyqir-generator/README.md [hunk 17]
```diff
@@ -1,22 +1,102 @@
-# pyqir-generator
+# pyqir_generator
 
-## Building and Testing
+The `pyqir_generator` package provides the ability to generate [QIR](https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR#quantum-intermediate-representation-qir) using a [builder](https://en.wikipedia.org/wiki/Builder_pattern) API.
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply pyqir-generator/README.md]: As per offline conversation, I took the liberty of applying the change.

File: .github/workflows/ci.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (bettinaheim) [reply .github/workflows/ci.yml:5]: As per offline conversation, I took the liberty of applying the change.

File: examples/generator/bernstein_vazirani.py [hunk 18]
```diff
@@ -0,0 +1,189 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply examples/generator/bernstein_vazirani.py:1]: will be a separate PR.

File: pyqir-jit/README.md [hunk 19]
```diff
@@ -1,22 +1,75 @@
...
 
-Alternatively, install tox and run the tests inside an isolated environment:
+## Limitations
 
-```shell
-tox -e py
-```
\ No newline at end of file
+- QIR entrypoint for JIT must be named `QuantumApplication__Run`
+- QIR must contain the defined runtime in [module.ll](../qirlib/src/module.ll)
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply pyqir-jit/README.md:82]: I added a comment that this is included if the pyqir_generator package is used to generate the QIR.

File: pyqir-generator/README.md [hunk 20]
```diff
@@ -1,22 +1,102 @@
...
 
-Alternatively, install tox and run the tests inside an isolated environment:
+## Limitations
 
-```shell
-tox -e py
-```
\ No newline at end of file
+- API is designed for automated translation, not as a circuit building library
+- Using qubit/register names in gate calls that haven't been defined will cause an error during generation.
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply pyqir-generator/README.md:101]: I added a line and added a link to a newly created issue.

File: pyqir-jit/README.md (hunk 7 above)
Comment (bettinaheim) [reply pyqir-jit/README.md:81]: I added a line that states that entry point arguments are not yet supported.

File: pyqir-jit/README.md (hunk 8 above)
Comment (bettinaheim) [reply pyqir-jit/README.md:11]: This will be done in a separate PR.

File: pyqir-generator/README.md [hunk 21]
```diff
@@ -1,22 +1,102 @@
...
 
-To build this package, first install `maturin`:
+"""
+Whether the two qubits should be entangled.
+    If "true", the two qubits are entangled such that the state of both
+    is one of the Bell states: | x 〉| y 〉 = 1/sqrt(2) [|0〉|0〉 + |1〉|1〉]
+    If "false", the two qubits are entangled such that the state is:
+    | x 〉| y 〉 = 1/4 [|0〉|0〉 + |0〉|1〉 + |1〉|0〉 + |1〉|1〉]
+"""
+entangle: bool = True
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply pyqir-generator/README.md]: I removed the boolean parameter, updated the superposition examples accordingly, and renamed it to bell_pair.py

**Review by bettinaheim — APPROVED** [review]
Approving with some additional work left for a separate PR.
